compiler/rustc_ast_lowering/src/item.rs

```diff
@@ -1254,11 +1254,13 @@ impl<'hir> LoweringContext<'_, 'hir> {
         coroutine_kind: Option<CoroutineKind>,
     ) -> (&'hir hir::Generics<'hir>, hir::FnSig<'hir>) {
         let header = self.lower_fn_header(sig.header);
+        // Don't pass along the user-provided constness of trait associated functions, we don't want to
+        // synthesize a host effect param for them. We reject `const` on them during AST validation.
+        let constness = if kind == FnDeclKind::Inherent { sig.header.constness } else { Const::No };
```

Comment:
This is strictly better than turning the `bug!("parent also has host effect param? […]")` into a `span_delayed_bug`. The latter can lead to weird extra diagnostics. E.g. if the constness found in the trait doesn't match with the one found in the trait impl, we would get diagnostics like ``method `f` has 0 const parameters but its trait declaration has 1 const parameter`` (and patching `compare_impl_item` to fix that would be a blatant hack).

Reply:
Can we not just do both (delay span bug, recover as `Const::No`?

Reply:
Sure